### PR TITLE
Create PR interstitial dialog refinements

### LINKED ISSUE #2265: Create Pull Request > Publish should show progress
1. Create a new branch.
1. Branch > Create Pull Request.
1. Click the Publish button.

Expected: the buttons should disable and a spinner should appear to indicate something is happening.

Actual: nothing happens until the publish is done.

### PR DESCRIPTION
**Pull request by niik**
I started out with the intention of just adding progress indication to the create PR interstitial dialog but got a bit carried away and did some refactoring, cleanup and style changes along the way.

# Before

image

image

# After

image

The dialog title as well as the content are rewritten to form an answerable question. The branch name has been emphasized using the new `Ref` component instead of `<b>` which matches what we use elsewhere (we'll gradually move other custom styles over to this new component). I've also broken up the text a bit for readability.

image

Here I've simply wired up the completion of the dispatcher promise to the `Dialog`'s `loading` prop which disables all buttons and adds the spinner up top.

image

Finally I made the same text changes in the `ahead` view of the dialog to have the text form an answerable question. Note that the `Cancel` button have changed here to `No`, meaning that if you click `No` here we won't push the branch but will still open the browser. Conceivably we could make this a `Yes` `No` `Cancel` dialog but since the `Create PR` action is a non-destructive one (just opening a web page) I didn't feel that was warranted.

Fixes #2265 

Sidenote, @donokuda: I wonder if we should play around with a dedicated question dialog in the same vein as #2277 with a question octicon offsetting the content. Might be worth thinking about. With it we could enforce (albeit only by documentation) that such dialogs should pose an answerable question.

### CODE REVIEW

File: app/src/ui/branches/PushBranchCommits.tsx [hunk 1]
```diff
@@ -16,34 +17,68 @@ interface IPushBranchCommitsProps {
...
+   */
+  readonly loading: boolean
+}
+
+/**
+ * Returns a string used for communicating the number of commits
+ * that will be pushed to the user. If only one commit is to be pushed
+ * we return the singular 'commit', if any other amount of commits
+ * are to be pushed we return the plural 'commits'. If the
+ * capitalize parameter is true we'll capitalize the 'c' in commit
```
Comment (joshaber) [inline app/src/ui/branches/PushBranchCommits.tsx]: This doesn't have a `capitalize` parameter anymore.

File: app/src/ui/branches/PushBranchCommits.tsx [hunk 2]
```diff
@@ -54,98 +89,98 @@ export class PushBranchCommits extends React.Component<
...
-    if (this.isPublish) {
-      return __DARWIN__
-        ? 'Publish Branch and Open Pull Request'
-        : 'Publish branch and open pull request'
+  private renderButtonGroup() {
+    if (renderPublishView(this.props.unPushedCommits)) {
+      return (
+        <ButtonGroup destructive={true}>
+          <Button type="submit">Cancel</Button>
+          <Button onClick={this.onPushOrPublishButtonClick}>Yes</Button>
```
Comment (joshaber) [inline app/src/ui/branches/PushBranchCommits.tsx]: On Mac they recommend against "Yes" as a button title and recommend using the action as the affirmative answer.

File: app/src/ui/branches/PushBranchCommits.tsx [hunk 3]
```diff
@@ -54,98 +89,98 @@ export class PushBranchCommits extends React.Component<
...
 
-    return __DARWIN__
-      ? 'Push Commits and Open Pull Request'
-      : 'Push commits and open pull request'
+    return (
+      <ButtonGroup destructive={true}>
+        <Button type="submit" onClick={this.onCreateWithoutPushButtonClick}>
+          No
+        </Button>
+        <Button onClick={this.onPushOrPublishButtonClick}>Yes</Button>
```
Comment (joshaber) [inline app/src/ui/branches/PushBranchCommits.tsx]: Same as above about "Yes"

File: app/src/ui/lib/ref.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * A simple style component used to mark up arbitrary references such as
+ * branches, commit shas, paths or other content which needs to be presented
+ * in an emphasized way and that benefit from fixed-width fonts.
+ *
+ * While the styling of the component _may_ differs depending on what context
+ * it appears in the general style is an inline-box with a suitable background
+ * color, using a fixed-width font.
+ */
+export class Ref extends React.Component<{}, {}> {
```
Comment (joshaber) [inline app/src/ui/lib/ref.tsx]: Nice 🤘

File: app/src/ui/lib/ref.tsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import * as React from 'react'
+
+/**
+ * A simple style component used to mark up arbitrary references such as
+ * branches, commit shas, paths or other content which needs to be presented
```
Comment (j-f1) [inline app/src/ui/lib/ref.tsx]: Nitpick: `SHAs`

File: app/styles/ui/_ref.scss [hunk 6]
```diff
@@ -0,0 +1,8 @@
+em.ref {
```
Comment (j-f1) [inline app/styles/ui/_ref.scss]: Maybe just `.ref` so the React component can be switched to another element if needed?

File: app/src/ui/branches/PushBranchCommits.tsx (hunk 2 above)
Comment (niik) [inline app/src/ui/branches/PushBranchCommits.tsx]: Hmm... I think the original dialog was much to wordy so I'm not keen on going back on this. I can see not only using `Yes` though. What would the guidelines say about doing something like `Yes, publish branch` and `Yes, push commits`?

**Comment by niik** [comment]
I tried out a new approach for the buttons. What I'm trying to achieve here that I think is important is that we ask an answerable question in these types of dialogs.

image

image

@joshaber how do you feel about that?

**Comment by joshaber** [comment]
It's better than the plain Yes/No, though:

>What I'm trying to achieve here that I think is important is that we ask an answerable question in these types of dialogs.

I agree, but conventionally these answers don't _actually_ say "yes" or "no." For example:

screen shot 2017-07-21 at 1 13 46 pm

**Comment by niik** [comment]
Alright, I want to get this in before my vacation so I'll just remove the Yes, No completely but I would like to look at platform use of `Yes` `No` buttons as I do suspect that they're a de-facto convention on Windows.

Updated images with the dialog refinements from #2277 brought in

image

image

**Pull request merged**